File: python_port/rag/impl/heuristic_query_writer.py
import re
from typing import List
from ..interfaces import QueryWriter, Intent

class HeuristicQueryWriter(QueryWriter):
    def __init__(self, stop_words: List[str]):
        self.stop_words = [word.lower() for word in stop_words]

    def write(self, question: str, intent: Intent) -> List[str]:
        # Remove non-alphanumeric characters (keeping Turkish chars and spaces)
        cleaned = re.sub(r"[^a-zA-Z0-9çğıöşüÇĞİÖŞÜ ]", "", question.lower())
        tokens = cleaned.split()
        
        terms = []
        for token in tokens:
            if token not in self.stop_words and token:
                terms.append(token)
        
        # Add extra words based on intent (Booster)
        if intent == Intent.StaffLookup:
            terms.extend(["staff", "advisor", "office"])
        
        # Return distinct terms while preserving order
        return list(dict.fromkeys(terms))

File: Iteration1CSE3063Proje/src/main/java/com/cse3063/rag/impl/HeuristicQueryWriter.java
package com.cse3063.rag.impl;

import com.cse3063.rag.interfaces.QueryWriter;
import com.cse3063.rag.interfaces.IntentDetector.Intent;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class HeuristicQueryWriter implements QueryWriter {
    private final List<String> stopWords;

    public HeuristicQueryWriter(List<String> stopWords) {
        this.stopWords = stopWords.stream().map(String::toLowerCase).collect(Collectors.toList());
    }

    @Override
    public List<String> write(String question, Intent intent) {
        String[] tokens = question.toLowerCase().replaceAll("[^a-zA-Z0-9çğıöşüÇĞİÖŞÜ ]", "").split("\\s+");
        List<String> terms = new ArrayList<>();

        for (String token : tokens) {
            if (!stopWords.contains(token) && !token.isEmpty()) {
                terms.add(token);
            }
        }

        // Add extra words based on intent (Booster)
        if (intent == Intent.StaffLookup) {
            terms.add("staff");
            terms.add("advisor");
            terms.add("office");
        }

        return terms.stream().distinct().collect(Collectors.toList());
    }
}
